### PR TITLE
Fix getdents syscall to handle seekdir/telldir

### LINKED ISSUE #8526: telldir/seekdir do not work
As far as I can tell we implement the getdents syscall (__syscall220 in library_syscall.js) but telldir always returns zero (since stream.position is always zero) and seekdir does nothing (since lseek doesn't seem to effect getdents as it should.

### PR DESCRIPTION
**Pull request by rianhunter**
Directory seeking was broken, this change makes syscall220() respect
the current directory offset.

Fixes #8526

### CODE REVIEW

File: system/lib/fetch/asmfs.cpp [hunk 1]
```diff
@@ -1910,7 +1910,7 @@ long __syscall220(int which, ...) // getdents64 (get directory entries 64-bit)
 	if (desc->file_pos <= file_pos)
 	{
 		de->d_ino = (ino_t)node; // TODO: Create inode numbers instead of using pointers
-		de->d_off = file_pos;
+		de->d_off = file_pos + sizeof(dirent);
```
Comment (sbc100) [inline system/lib/fetch/asmfs.cpp]: It looks like this code is for the first entry read , so shouldn't the d_off for that entry be 0?     There it looks like d_off is being set to the end of the entry?   Is this part of the change tested somewhere?

File: src/library_syscall.js [hunk 2]
```diff
@@ -1081,11 +1081,17 @@ var SyscallsLibrary = {
...
-    while (stream.getdents.length > 0 && pos + {{{ C_STRUCTS.dirent.__size__ }}} <= count) {
+    var off = FS.llseek(stream, 0, {{{ cDefine('SEEK_CUR') }}});
+
+    var idx = Math.floor(off / struct_size);
+
+    while (idx < stream.getdents.length && pos + struct_size <= count) {
       var id;
       var type;
-      var name = stream.getdents.pop();
+      var name = stream.getdents[stream.getdents.length - 1 - idx];
```
Comment (sbc100) [inline src/library_syscall.js]: Hmm. .are we reading these backwards?  Why not just `stream.getdents[idx]`?

File: system/lib/fetch/asmfs.cpp (hunk 1 above)
Comment (rianhunter) [reply system/lib/fetch/asmfs.cpp]: This is tested implicitly in test_readdir.c by using telldir() but not by using d_off. d_off indeed points to the directory entry after. I can add that test as well!

File: src/library_syscall.js (hunk 2 above)
Comment (rianhunter) [reply src/library_syscall.js]: The old code iterated backwards by using `.pop()` this just preserves that behavior. I can put a comment there to make that clear.

Comment (sbc100) [reply src/library_syscall.js]: if we switch to forward iteration presumably we can avoid the change in `lib/fetch/asmfs.cpp`?   Do we have any tests that break?  (i.e. rely on the order iteration?)

Comment (rianhunter) [reply src/library_syscall.js]: The `d_off` change in `asmfs.cpp` is not related to iteration order, that's just how `readdir()` works. The tests that depend on iteration order are `other.test_mkdir_silly` and `other.test_readdir_r_silly`

Comment (rianhunter) [reply src/library_syscall.js]: (I'm open to changing those tests for the purposes of changing iteration order)

File: system/lib/fetch/asmfs.cpp (hunk 1 above)
Comment (sbc100) [reply system/lib/fetch/asmfs.cpp]: I think d_off is supposed to the point the beginning of the entry, not the end?  

I can see that this change would make sense when iterating backwards, but if iterate forwards than wouldn't we want d_off to point the start of the current record?  (In which case this change is not needed?)

File: src/library_syscall.js (hunk 2 above)
Comment (sbc100) [reply src/library_syscall.js]: I think its worth a try.

File: system/lib/fetch/asmfs.cpp (hunk 1 above)
Comment (rianhunter) [reply system/lib/fetch/asmfs.cpp]: `d_off` should be the same value as `telldir()` after the call to `readdir()`, this isn't related to iterating forwards or backwards.

Comment (rianhunter) [reply system/lib/fetch/asmfs.cpp]: See https://linux.die.net/man/2/getdents64

Comment (sbc100) [reply system/lib/fetch/asmfs.cpp]: Ah!  I see.

**Comment by rianhunter** [comment]
Gentle ping

**Comment by sbc100** [comment]
Can you add yourself the AUTHORS file if you are not already there?

**Comment by rianhunter** [comment]
Already on AUTHORS! https://github.com/emscripten-core/emscripten/blob/07de93c796f601507dbd56d2c1a4b0dcbc4424f3/AUTHORS#L381

**Pull request merged**